### ISSUE TITLE
rc1 INCOMPLETE_CHUNKED_ENCODING when using nginx

### ISSUE THREAD

**Issue by staff0rd** [aspnet/KestrelHttpServer#341]
Opened this [here](/jwilder/nginx-proxy/issues/284) also, but I'm getting the feeling this more specific to kestrel rc1.

I have a docker container that has an aspnet5-rc1 application inside, listening on port 5000. If I make a request directly to that port, the response is 200 and quick.

If I make a request via [nginx](/jwilder/nginx-proxy), then the response data comes back immediately (200), but does not terminate and rather about a minute later the response fails with; (in chrome) ERR_INCOMPLETE_CHUNKED_ENCODING.

Error does not occur when using nginx with aspnet-beta7.  Headers for both aspnet-beta7 and aspnet-rc1 are the same;

~~~
HTTP/1.1 200 OK
Server: nginx/1.9.6
Date: Sun, 08 Nov 2015 04:38:28 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Frame-Options: SAMEORIGIN
~~~

**Comment by Tratcher** [comment]
I have a local repro that looks similar. If I throw an exception after the first response write then there is no chunked terminator and the connection is closed gracefully rather than reset.
```
            app.Run(async (context) =>
            {
                await context.Response.WriteAsync("hello world");
                throw new Exception("Test exception");
            });
```
The behavior is even stranger behind HttpPlatformHandler. It appears to suppress the chunking error and re-chunk the response including a terminator. Verify if it behaves correctly after kestrel changes to a reset.

**Comment by staff0rd** [comment]
@Tratcher any hints as to why I'm seeing this behind nginx reverse proxy but not when accessing it directly?

**Comment by Tratcher** [comment]
@staff0rd That's odd. In my case the reverse proxy (HttpPlatformHandler) hid the behavior by re-chunking the response.

You just posted the headers, did any of the body make it through? Or did you do an explicit Flush to send the headers before an exception occurred?

**Comment by staff0rd** [comment]
@Tratcher if you browse to http://test.atqu.in you can see the issue occur.  If you browse to http://test.atqu.in:5000 you'll reach the same site directly (without reverse proxy) and you'll see the issue does not occur.  As you'll see, the whole site is is delivered both times, but the connection stays open on the reverse-proxy for about a minute until it finally dies with INCOMPLETE_CHUNKED_ENCODING.

The code is [here](/staff0rd/TinyMvc) and is just a fresh MVC 6 beta8 project updated to rc1, so no extra flushes occur, nor are any exceptions thrown.

**Comment by Tratcher** [comment]
@staff0rd Let's re-test your scenario after we get the identified bug fixed.

**Comment by cesarbs** [comment]
The problem here is the same as what we're facing in https://github.com/aspnet/KestrelHttpServer/issues/368: we're not sending an RST when we expect to (`ConnectionControl.End(ProduceEndType.SocketDisconnect)`).

Libuv currently does not expose the underlying socket enough for us to force an RST. They did have an API in the past that would allow us to send an RST (albeit Unix-only), but that is not there anymore: https://github.com/joyent/libuv/commit/8b213a9cb6464dd5a4368cc71bd220d9d83f97ea

Basically what needs to be done is to set `SO_LINGER` on the socket without a timeout of 0 (which you can see in the commite linked above).

I'll try to find out why they removed that from their code.

**Comment by cesarbs** [comment]
Forget my previous comment :smile:

nginx makes HTTP/1.0 requests when acting as a proxy. There is a known issue for `Connection: close` requests (which HTTP/1.0 requests implicitly are): https://github.com/aspnet/KestrelHttpServer/issues/406

There's an easy fix for putting Kestrel behind an nginx reverse proxy: set `proxy_http_version 1.1;` to have nginx send HTTP/1.1 requests.

**Comment by staff0rd** [comment]
[This](https://github.com/jwilder/nginx-proxy/blob/master/nginx.tmpl#L43-L51) is already configuring HTTP/1.1.  Looking at the response headers of [test.atqu.in](http://test.atqu.in) also confirms that HTTP/1.1 is being used - same headers as in first post above.

**Comment by cesarbs** [comment]
Can you try a different version of nginx? I repro'd your issue on 1.8 and found that using that setting fixed the problem. Then I switched to 1.9.6 and on that specific version the requests seem to take a long time to be served (but I didn't see the original issue), even when running on localhost. I don't see that happening on 1.9.7 though.

**Comment by staff0rd** [comment]
[Updated](http://test.atqu.in), same symptoms - 1.3min response time.

Response headers;
~~~
HTTP/1.1 200 OK
Server: nginx/1.9.7
Date: Thu, 19 Nov 2015 23:49:28 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
~~~

**Comment by cesarbs** [comment]
I can hit that URL just fine. Wasn't that the URL where you were seeing the error?

**Comment by staff0rd** [comment]
Seriously?  I just tried chrome + firefox + edge.  Same symptoms on all - page immediately pops up, but loading doesn't stop for 1 min.  However, the page imediately pops up *and* stops loading if I go around nginx. (http://test.atqu.in:5000)

**Comment by cesarbs** [comment]
Ohh, sorry, I didn't notice it was still loading :disappointed:

What addresses are you binding to? I figured out why 1.9.6 was taking so long to forward the request to Kestrel. I was binding Kestrel to my IPv6 loopback interface only, but nginx was first trying to forward the request to the IPv4 interface. It took a minute to time out and try the IPv6 interface. Maybe something similar is happening in your setup?

**Comment by staff0rd** [comment]
Do you mean [this](https://github.com/staff0rd/TinyMvc/blob/master/project.json#L20) or something on the nginx side?

**Comment by cesarbs** [comment]
Yep, that. Although you're already binding to an IPv4 interface... Can you try binding to localhost or * and see if that makes a difference?

**Comment by staff0rd** [comment]
Its hosted in a docker container so I don't think that binding to localhost will work.  What's the syntax for \*,  `http://*:5000`?

**Comment by cesarbs** [comment]
That's correct.

**Comment by staff0rd** [comment]
Nah, same result;

~~~
Failed to load resource: net::ERR_INCOMPLETE_CHUNKED_ENCODING
~~~

**Comment by cesarbs** [comment]
Ok, I've managed to repro your issue and it's in fact the same issue linked above - https://github.com/aspnet/KestrelHttpServer/issues/406.

For some reason nginx sends a `Connection: close` header to Kestrel. The nginx-proxy template actually erases any `Connection` headers in the request, but that's not really the issue because for HTTP/1.1 no `Connection` == `Connection: keep-alive`.

When I added

    proxy_set_header Connection keep-alive;

to my nginx.conf the issue was gone. Can you try doing that?

**Comment by staff0rd** [comment]
nice one @cesarbs, that indeed [fixes it](http://test.atqu.in).  I don't understand how/why that fixes though, nor why nginx-proxy removes it, I guess I should take it up with them.

**Comment by cesarbs** [comment]
Closing this since we've found a workaround and we know that https://github.com/aspnet/KestrelHttpServer/issues/406 will eventually fix the problem.

**Comment by cesarbs** [comment]
@staff0rd The real issue is #406, the `Connection: close` sent by nginx hits that issue. Setting `Connection: keep-alive` avoids the buggy code path.

**Comment by tugberkugurlu** [comment]
had the same issue, luckily has been pointed to hear before losing all my hairs :smile:  http://stackoverflow.com/q/34821881/463785

**Comment by tobz** [comment]
Just for anyone that stumbles on this issue: I'm also seeing this behavior when using the webpack-dev-server/hot reload setup proxying to Kestrel.  Talking straight to Kestrel works as intended.  I haven't yet investigated if a temporary workaround exists or not.

**Comment by tobz** [comment]
Spent a little time looking at it, and forcing `node-http-proxy` to send `Connection: keep-alive` to Kestrel seems to do the trick.  Here's the `proxy` section of my webpack-dev-server configuration:

    devServer: {
        proxy: {
            '*': {
                target: 'http://localhost:5000/',
                secure: false,
                changeOrigin: true,
                headers: {
                    'Connection': 'keep-alive'
                }
          },
    },

**Comment by paralin** [comment]
Can this be fixed for real?

Nginx sends `Connection: close` because when proxying requests it's best to close the connection immediately and not keep it alive.

**Comment by benaadams** [comment]
@paralin this is fixed in RC2

**Comment by paralin** [comment]
@benaadams Any way a fix for this can be made for rc1? I don't think switching the entire stack to the AspCore dev versions is wise but I can apply a single patch to rc1.

**Comment by davidfowl** [comment]
See https://github.com/aspnet/KestrelHttpServer/issues/341#issuecomment-158247405 for a workaround:

```
proxy_set_header Connection keep-alive;
```

**Comment by amcdnl** [comment]
@benaadams I'm still experiencing this in RC2

**Comment by cesarbs** [comment]
#636 #734 seem to be the same as this one (actually #636 is about BrowserSync). Let's center the discussion here, since this was the original issue.

@amcdnl Which version of nginx are you using, and on which OS?

**Comment by halter73** [comment]
This issue is related to Kestrel not sending the chunked response suffix. This was fixed by e4fd91b which closed this issue and #406.

Unless people are still seeing INCOMPLETE_CHUNKED_ENCODING in chrome after the nginx timeout, I would close this.

**Comment by amcdnl** [comment]
@cesarbs  I updated the details in https://github.com/aspnet/KestrelHttpServer/issues/636 to reflect your questions.

**Comment by mikeharder** [comment]
@amcdnl: Can we close this issue, or are you still seeing INCOMPLETE_CHUNKED_ENCODING with RC2?

**Comment by amcdnl** [comment]
@mikeharder Good to go.

**Comment by jmbalanag** [comment]
Im currently encoutering this issue:

"Microsoft.AspNetCore.Server.Kestrel": "1.0.0",

**Comment by davidfowl** [comment]
@jmbalanag can you provide more details about your setup and provide specific repro steps?

**Comment by jmbalanag** [comment]
This is the problem code:
   [Route("api/lib_fund_source")]
        public ActionResult GetFundSource()
        {
            return Json(db.lib_fund_source.Where(x => x.is_active == true).Select(x => new { Id = x.fund_source_id, Name = x.name }));

        }

if i remove, Where(x => x.is_active == true) it works just fine. I

**Comment by muratg** [comment]
@jmbalang Yeah, doesn't look like a Kestrel issue
